File: controllers/authController.js
/**
 * ============================================
 * 🔐 Auth Controller - نظام إدهام
 * Edham Logistics - Authentication Controller
 * ============================================
 */

const User = require('../models/User');
const { MESSAGES, HTTP_STATUS } = require('../config/constants');
const logger = require('../utils/logger');
const AuthService = require('../modules/auth/auth.service');

class AuthController {
  /**
   * Register new user
   */
  static async register(req, res) {
    try {
      const result = await AuthService.register(req.body);
      if (!result.ok) {
        return res.status(result.statusCode).json({ success: false, message: result.message });
      }
      logger.success('User registered', { email: result.user.email, role: result.user.role });
      res.status(result.statusCode).json({
        success: true,
        message: result.message,
        token: result.token,
        user: result.user
      });
    } catch (error) {
      logger.error('Registration error', error);
      res.status(HTTP_STATUS.INTERNAL_ERROR).json({
        success: false,
        message: MESSAGES.ERROR,
        error: error.message
      });
    }
  }

  /**
   * Login user
   */
  static async login(req, res) {
    try {
      const result = await AuthService.login(req.body);
      if (!result.ok) {
        return res.status(result.statusCode).json({ success: false, message: result.message });
      }
      logger.success('User logged in', { email: result.user.email, role: result.user.role });
      res.status(result.statusCode).json({
        success: true,
        message: result.message,
        token: result.token,
        user: result.user
      });
    } catch (error) {
      logger.error('Login error', error);
      res.status(HTTP_STATUS.INTERNAL_ERROR).json({
        success: false,
        message: MESSAGES.ERROR
      });
    }
  }

  /**
   * Get current user
   */
  static async getCurrentUser(req, res) {
    try {
      const user = await User.findById(req.user._id).select('-password');
      res.json({
        success: true,
        data: user
      });
    } catch (error) {
      res.status(HTTP_STATUS.INTERNAL_ERROR).json({
        success: false,
        message: MESSAGES.ERROR
      });
    }
  }

  /**
   * Update user profile
   */
  static async updateProfile(req, res) {
    try {
      const { name, phone, address, city, avatar } = req.body;

      const user = await User.findByIdAndUpdate(
        req.user._id,
        { name, phone, address, city, avatar, updatedAt: new Date() },
        { new: true, runValidators: true }
      ).select('-password');

      logger.success('Profile updated', { userId: req.user._id });

      res.json({
        success: true,
        message: MESSAGES.UPDATED,
        data: user
      });
    } catch (error) {
      res.status(HTTP_STATUS.INTERNAL_ERROR).json({
        success: false,
        message: MESSAGES.ERROR
      });
    }
  }

  /**
   * Change password
   */
  static async changePassword(req, res) {
    try {
      const { currentPassword, newPassword } = req.body;

      const user = await User.findById(req.user._id).select('+password');

      const isValid = await user.comparePassword(currentPassword);
      if (!isValid) {
        return res.status(HTTP_STATUS.UNAUTHORIZED).json({
          success: false,
          message: 'Current password is incorrect'
        });
      }

      user.password = newPassword;
      await user.save();

      logger.success('Password changed', { userId: req.user._id });

      res.json({
        success: true,
        message: 'Password changed successfully'
      });
    } catch (error) {
      res.status(HTTP_STATUS.INTERNAL_ERROR).json({
        success: false,
        message: MESSAGES.ERROR
      });
    }
  }
}

module.exports = AuthController;

File: mobile-native-android/backend-api/src/main/java/com/edham/logistics/controller/AuthController.java
package com.edham.logistics.controller;

import com.edham.logistics.dto.LoginRequest;
import com.edham.logistics.dto.LoginResponse;
import com.edham.logistics.dto.RefreshTokenRequest;
import com.edham.logistics.dto.RegisterRequest;
import com.edham.logistics.service.AuthService;
import com.edham.logistics.util.ApiResponse;
import jakarta.validation.Valid;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import jakarta.servlet.http.HttpServletRequest;

/**
 * Authentication Controller
 * Handles user authentication, registration, and token management
 */
@RestController
@RequestMapping("/api/v1/auth")
@CrossOrigin(origins = "*", maxAge = 3600)
public class AuthController {

    @Autowired
    private AuthService authService;

    /**
     * User login
     */
    @PostMapping("/login")
    public ResponseEntity<ApiResponse<LoginResponse>> login(
            @Valid @RequestBody LoginRequest loginRequest,
            HttpServletRequest request) {
        
        try {
            String clientIp = getClientIpAddress(request);
            String userAgent = request.getHeader("User-Agent");
            
            LoginResponse response = authService.authenticate(
                loginRequest.getEmail(),
                loginRequest.getPassword(),
                clientIp,
                userAgent
            );
            
            return ResponseEntity.ok(
                ApiResponse.<LoginResponse>builder()
                    .success(true)
                    .message("Login successful")
                    .data(response)
                    .build()
            );
            
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED)
                .body(ApiResponse.<LoginResponse>builder()
                    .success(false)
                    .message("Invalid credentials: " + e.getMessage())
                    .build()
                );
        }
    }

    /**
     * User registration
     */
    @PostMapping("/register")
    public ResponseEntity<ApiResponse<LoginResponse>> register(
            @Valid @RequestBody RegisterRequest registerRequest,
            HttpServletRequest request) {
        
        try {
            String clientIp = getClientIpAddress(request);
            String userAgent = request.getHeader("User-Agent");
            
            LoginResponse response = authService.register(
                registerRequest,
                clientIp,
                userAgent
            );
            
            return ResponseEntity.status(HttpStatus.CREATED)
                .body(ApiResponse.<LoginResponse>builder()
                    .success(true)
                    .message("Registration successful")
                    .data(response)
                    .build()
                );
            
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.BAD_REQUEST)
                .body(ApiResponse.<LoginResponse>builder()
                    .success(false)
                    .message("Registration failed: " + e.getMessage())
                    .build()
                );
        }
    }

    /**
     * Refresh access token
     */
    @PostMapping("/refresh")
    public ResponseEntity<ApiResponse<LoginResponse>> refreshToken(
            @Valid @RequestBody RefreshTokenRequest refreshRequest,
            HttpServletRequest request) {
        
        try {
            String clientIp = getClientIpAddress(request);
            
            LoginResponse response = authService.refreshToken(
                refreshRequest.getRefreshToken(),
                clientIp
            );
            
            return ResponseEntity.ok(
                ApiResponse.<LoginResponse>builder()
                    .success(true)
                    .message("Token refreshed successfully")
                    .data(response)
                    .build()
            );
            
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED)
                .body(ApiResponse.<LoginResponse>builder()
                    .success(false)
                    .message("Token refresh failed: " + e.getMessage())
                    .build()
                );
        }
    }

    /**
     * User logout
     */
    @PostMapping("/logout")
    public ResponseEntity<ApiResponse<Void>> logout(
            @RequestHeader("Authorization") String authorization,
            HttpServletRequest request) {
        
        try {
            String token = extractTokenFromHeader(authorization);
            String clientIp = getClientIpAddress(request);
            
            authService.logout(token, clientIp);
            
            return ResponseEntity.ok(
                ApiResponse.<Void>builder()
                    .success(true)
                    .message("Logout successful")
                    .build()
            );
            
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.BAD_REQUEST)
                .body(ApiResponse.<Void>builder()
                    .success(false)
                    .message("Logout failed: " + e.getMessage())
                    .build()
                );
        }
    }

    /**
     * Validate token
     */
    @GetMapping("/validate")
    public ResponseEntity<ApiResponse<Boolean>> validateToken(
            @RequestHeader("Authorization") String authorization) {
        
        try {
            String token = extractTokenFromHeader(authorization);
            boolean isValid = authService.validateToken(token);
            
            return ResponseEntity.ok(
                ApiResponse.<Boolean>builder()
                    .success(true)
                    .message("Token validation completed")
                    .data(isValid)
                    .build()
            );
            
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED)
                .body(ApiResponse.<Boolean>builder()
                    .success(false)
                    .message("Token validation failed: " + e.getMessage())
                    .data(false)
                    .build()
                );
        }
    }

    /**
     * Request password reset
     */
    @PostMapping("/forgot-password")
    public ResponseEntity<ApiResponse<Void>> forgotPassword(
            @RequestParam String email,
            HttpServletRequest request) {
        
        try {
            String clientIp = getClientIpAddress(request);
            
            authService.forgotPassword(email, clientIp);
            
            return ResponseEntity.ok(
                ApiResponse.<Void>builder()
                    .success(true)
                    .message("Password reset email sent")
                    .build()
            );
            
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.BAD_REQUEST)
                .body(ApiResponse.<Void>builder()
                    .success(false)
                    .message("Password reset failed: " + e.getMessage())
                    .build()
                );
        }
    }

    /**
     * Reset password with token
     */
    @PostMapping("/reset-password")
    public ResponseEntity<ApiResponse<Void>> resetPassword(
            @RequestParam String token,
            @RequestParam String newPassword,
            HttpServletRequest request) {
        
        try {
            String clientIp = getClientIpAddress(request);
            
            authService.resetPassword(token, newPassword, clientIp);
            
            return ResponseEntity.ok(
                ApiResponse.<Void>builder()
                    .success(true)
                    .message("Password reset successful")
                    .build()
            );
            
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.BAD_REQUEST)
                .body(ApiResponse.<Void>builder()
                    .success(false)
                    .message("Password reset failed: " + e.getMessage())
                    .build()
                );
        }
    }

    /**
     * Change password (authenticated user)
     */
    @PostMapping("/change-password")
    public ResponseEntity<ApiResponse<Void>> changePassword(
            @RequestParam String currentPassword,
            @RequestParam String newPassword,
            @RequestHeader("Authorization") String authorization,
            HttpServletRequest request) {
        
        try {
            String token = extractTokenFromHeader(authorization);
            String clientIp = getClientIpAddress(request);
            
            authService.changePassword(token, currentPassword, newPassword, clientIp);
            
            return ResponseEntity.ok(
                ApiResponse.<Void>builder()
                    .success(true)
                    .message("Password changed successfully")
                    .build()
            );
            
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.BAD_REQUEST)
                .body(ApiResponse.<Void>builder()
                    .success(false)
                    .message("Password change failed: " + e.getMessage())
                    .build()
                );
        }
    }

    /**
     * Get current user info
     */
    @GetMapping("/me")
    public ResponseEntity<ApiResponse<Object>> getCurrentUser(
            @RequestHeader("Authorization") String authorization) {
        
        try {
            String token = extractTokenFromHeader(authorization);
            Object userInfo = authService.getCurrentUser(token);
            
            return ResponseEntity.ok(
                ApiResponse.<Object>builder()
                    .success(true)
                    .message("User info retrieved successfully")
                    .data(userInfo)
                    .build()
            );
            
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED)
                .body(ApiResponse.<Object>builder()
                    .success(false)
                    .message("Failed to get user info: " + e.getMessage())
                    .build()
                );
        }
    }

    /**
     * Extract token from Authorization header
     */
    private String extractTokenFromHeader(String authorization) {
        if (authorization != null && authorization.startsWith("Bearer ")) {
            return authorization.substring(7);
        }
        throw new IllegalArgumentException("Invalid authorization header");
    }

    /**
     * Get client IP address
     */
    private String getClientIpAddress(HttpServletRequest request) {
        String xForwardedFor = request.getHeader("X-Forwarded-For");
        if (xForwardedFor != null && !xForwardedFor.isEmpty()) {
            return xForwardedFor.split(",")[0].trim();
        }
        
        String xRealIp = request.getHeader("X-Real-IP");
        if (xRealIp != null && !xRealIp.isEmpty()) {
            return xRealIp;
        }
        
        return request.getRemoteAddr();
    }
}
